lightning/src/ln/chan_utils.rs

```diff
@@ -550,11 +556,436 @@ pub fn build_htlc_transaction(prev_hash: &Txid, feerate_per_kw: u32, contest_del
+		// Sort outputs and populate output indices while keeping track of the auxiliary data
+		let mut txouts = Self::do_build_outputs(&keys, to_broadcaster_value_sat, to_countersignatory_value_sat, &htlcs, aux, channel_parameters, &secp_ctx).unwrap();
+		let mut result_htlcs_with_aux = Vec::new();
+		let mut htlcs = Vec::new();
+		for (idx, mut out) in txouts.drain(..).enumerate() {
+			if let Some(mut htlc) = (out.1).1.take() {
+				htlc.0.transaction_output_index = Some(idx as u32);
+				result_htlcs_with_aux.push((htlc.0.clone(), htlc.1));
+				htlcs.push(htlc.0);
+			}
+		}
+
+		let info = CommitmentTransaction {
+			commitment_number,
+			to_broadcaster_value_sat,
+			to_countersignatory_value_sat,
+			feerate_per_kw,
+			htlcs,
+			keys,
+		};
+		(info, result_htlcs_with_aux)
+	}
+
+	pub(crate) fn calculate_txid<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Txid {
+		let tx = self.build(channel_parameters, secp_ctx).unwrap().0;
+		tx.txid()
+	}
+
+	/// Build the Bitcoin transaction
+	pub fn build<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Result<(bitcoin::Transaction, Vec<Script>), ()> {
+		let (obscured_commitment_transaction_number, txins) = self.build_inputs(channel_parameters);
+
+		let mut txouts = self.build_outputs(channel_parameters, secp_ctx)?;
+
+		let mut outputs = Vec::with_capacity(txouts.len());
+		let mut scripts = Vec::with_capacity(txouts.len());
+		for (txout, script) in txouts.drain(..) {
+			outputs.push(txout);
+			scripts.push(script);
+		}
+
+		Ok((
+			Transaction {
+				version: 2,
+				lock_time: ((0x20 as u32) << 8 * 3) | ((obscured_commitment_transaction_number & 0xffffffu64) as u32),
+				input: txins,
+				output: outputs,
+			},
+			scripts,
+		))
+	}
+
+	fn build_outputs<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Result<Vec<(TxOut, Script)>, ()> {
+		let mut aux = Vec::with_capacity(self.htlcs.len());
+		aux.resize(self.htlcs.len(), ());
+		let mut txouts = Self::do_build_outputs(&self.keys, self.to_broadcaster_value_sat, self.to_countersignatory_value_sat, &self.htlcs, aux, channel_parameters, secp_ctx)?;
+		let outs = txouts.drain(..).map(|(out, (s, _))| (out, s)).collect();
+		Ok(outs)
+	}
+
+	fn do_build_outputs<T, S: secp256k1::Signing + secp256k1::Verification>(keys: &TxCreationKeys, to_broadcaster_value_sat: u64, to_countersignatory_value_sat: u64, htlcs: &Vec<HTLCOutputInCommitment>, aux: Vec<T>, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<S>) -> Result<Vec<(TxOut, (Script, Option<(HTLCOutputInCommitment, T)>))>, ()> {
```

Comment:
I know these methods are private, but with `build_outputs` and `do_build_outputs` sounding pretty much identical, could we add some comments to expand on what their differences? Renaming one of the methods might also be helpful.

Additionally, `do_build_outputs` takes an argument whose name is simply aux. I think that is also something where a comment explaining that it's simply an HTLC iteration crutch would help.

Reply:
will do!